### ISSUE TITLE
Simplify `read_only` attributes

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#566]
Right now we have both `project.read_only` and `dataset.read_only` attributes. None of the two is used, but we discussed in the past how at some point a user may want to disable updates to a certain object.

Do we really need this attribute for both projects and datasets? My first guess would be that we can drop it from the dataset table, and make everything a bit simpler.

**Comment by jluethi** [comment]
For the input datasets: Is this just a flag to say that an entry in the database shouldn't be changed or that Fractal shouldn't change the underlying data?

I always understood it (for the dataset) as: We set a flag whether Fractal is allowed to change files in a given folder. And Fractal only modifies the files for the output dataset, it never modifies the files for the input dataset when we use e.g. Yokogawa images as the input dataset.

**Comment by tcompa** [comment]
My bad, there is indeed a use of `dataset.read_only` (we never touch it, and I had forgotten about it).

It is used in the workflow-execution endpoint, in the heuristic to define the output dataset of a workflow, when the user did not specify it. The corresponding part of the endpoint looks like
```python
    if apply_workflow.output_dataset_id:
        output_dataset = ...
    else:
        try:
            output_dataset = await auto_output_dataset(
                project=project, input_dataset=input_dataset, workflow=workflow
            )
        except Exception as e:
            ...
```
and the function is
```python
async def auto_output_dataset(
    *,
    project: Project,
    input_dataset: Dataset,
    workflow: Workflow,
    overwrite_input: bool = False,
) -> Dataset:
    """
    Determine the output dataset if it was not provided explicitly

    Only datasets containing exactly one path can be used as output.

    Note: This routine is still a stub.

    Args:
        project:
            The project that contains the input and output datasets.
        input_dataset:
            The input dataset.
        workflow:
            The workflow to be applied to the input dataset.
        overwrite_input:
            Whether it is allowed to overwrite the input dataset with the
            output data.

    Raises:
        ValueError: If the input dataset is to be overwritten and it provides
                    more than one path.

    Returns:
        output_dataset:
            the output dataset
    """
    if overwrite_input and not input_dataset.read_only:
        input_paths = input_dataset.paths
        if len(input_paths) != 1:
            raise ValueError(
                "Cannot determine output dataset "
                "with more than one input path."
            )
        output_dataset = input_dataset
    else:
        raise NotImplementedError(
            "Cannot determine ouput dataset with "
            f"{overwrite_input=} and {input_dataset.read_only=}"
        )

    return output_dataset
```

Therefore removing `dataset.read_only` only makes sense if we think that this heuristic is not needed.

> For the input datasets: Is this just a flag to say that an entry in the database shouldn't be changed or that Fractal shouldn't change the underlying data?

The partial answer is that because this flag is used when defining whether an input dataset can also be used as an output dataset, then it helps preventing erroneous overwriting of files.
However, fractal-server has zero knowledge of what tasks do internally; for this reason, nothing prevents the task from modifying or removing the content of `input_paths` on disk. This becomes obvious when you think about the `delete_input` argument of `yokogawa_to_ome_zarr.py`.

TL;DR The only guarantee that comes from setting a dataset as read-only is that its database entry won't be modified. There is no guarantee about the files.

**Comment by tcompa** [comment]
Side note: the `overwrite_input` attribute of the workflow-apply payload is currently not even passed to the `auto_output_dataset` function (that's not on purpose, that's a bug), and so for the moment all this automatic dataset selection has never been used by anyone.

We'd be in favor of just fully removing that function and that feature. That is, we'd rather make the `output_dataset_id` a required parameter (rather than optional). This function doesn't seem sufficiently robust or useful to justify the additional complexity.
One would always be able to use the same dataset ID for both input and output, and in that case we'll raise an error if the dataset is read-only.

@jluethi @gusqgm any feedback on this?

**Comment by jluethi** [comment]
> Side note: the overwrite_input attribute of the workflow-apply payload is currently not even passed to the auto_output_dataset function (that's not on purpose, that's a bug), and so for the moment all this automatic dataset selection has never been used by anyone.
We'd be in favor of just fully removing that function and that feature. That is, we'd rather make the output_dataset_id a required parameter (rather than optional). This function doesn't seem sufficiently robust or useful to justify the additional complexity.
One would always be able to use the same dataset ID for both input and output, and in that case we'll raise an error if the dataset is read-only.

That seems reasonable to me. Actively setting the output dataset makes sense.

We should then do the work on the task side to ensure that tasks write to the output_dataset, not the input_dataset. If the 2 are the same and the read_only flag is set to true, Fractal server should stop it. If they are the same and the dataset is not read-only, it should run. If they are different, the question is just whether the output dataset is read-only or not.

**Comment by tcompa** [comment]
> We should then do the work on the task side to ensure that tasks write to the output_dataset, not the input_dataset.

Agreed.
To be clear, this will always be a risk. If the task says `subprocess.run("rm -r /data/homes/myself")`, that cannot be prevented on the fractal-server side.

> If the 2 are the same and the read_only flag is set to true, Fractal server should stop it. If they are the same and the dataset is not read-only, it should run. If they are different, the question is just whether the output dataset is read-only or not.

I think this boils down to "if the output dataset is read-only, then raise an error and stop"

**Comment by jluethi** [comment]
Agreed.

> I think this boils down to "if the output dataset is read-only, then raise an error and stop"

Agreed. If there is a chain of tasks in a workflow, only the first task is input to output, right? All other tasks then are output to output, right? At least in our current workflow logic where a user couldn't change the output dataset mid-workflow (see e.g. 2D to 3D discussions, though those are and probably would remain within the same dataset)

**Comment by tcompa** [comment]
> If there is a chain of tasks in a workflow, only the first task is input to output, right? All other tasks then are output to output, right?

That's logically how we use this, but that it is not part of the implementation. We only have datasets in the database for the input of the first task and for the output of the last task. These two are the ones identified via input_dataset_id and output_dataset_id.

The intermediate states are not stored in the database, for the usual reason (we'd rather not have database sessions handled as part of the backend runner, because it would mix different components).

We'll have to review this structure for the "run from A to B" feature, and that's also one reason for removing small components which complicate things (like automatic detection of output dataset).

**Comment by tcompa** [comment]
A last check was missing here, and currently fixed with [aedbf38](https://github.com/fractal-analytics-platform/fractal-server/commit/aedbf389a9032419a385f8896e23cbd89e3c6cee).
